overlay.go

```diff
@@ -331,3 +338,331 @@ func isZero(v reflect.Value) bool {
+				if err != nil {
+					return nil, errors.Annotatef(err, "processing option %q for application %q", k, app)
+				}
+				if changed {
+					appData.Options[k] = newV
+				}
+			}
+
+			for k, v := range appData.Annotations {
+				newV, changed, err := resolveIncludes(incResolver, v)
+				if err != nil {
+					return nil, errors.Annotatef(err, "processing annotation %q for application %q", k, app)
+				}
+				if changed {
+					appData.Annotations[k] = newV
+				}
+			}
+		}
+
+		for machine, machineData := range base.Data.Machines {
+			if machineData == nil {
+				continue
+			}
+
+			for k, v := range machineData.Annotations {
+				newV, changed, err := resolveIncludes(incResolver, v)
+				if err != nil {
+					return nil, errors.Annotatef(err, "processing annotation %q for machine %q", k, machine)
+				}
+				if changed {
+					machineData.Annotations[k] = newV
+				}
+			}
+		}
+	}
+
+	return base.Data, nil
+}
+
+func applyOverlay(base, overlay *BundleDataPart) error {
+	if !overlay.PresenceMap.fieldPresent("applications") && len(overlay.Data.Applications) > 0 {
+		return errors.Errorf("bundle overlay file used deprecated 'services' key, this is not valid for bundle overlay files")
+	}
+
+	// Merge applications
+	if len(overlay.Data.Applications) != 0 {
+		if base.Data.Applications == nil {
+			base.Data.Applications = make(map[string]*ApplicationSpec, len(overlay.Data.Applications))
+		}
+		fpm := overlay.PresenceMap.forField("applications")
+		for srcAppName, srcAppSpec := range overlay.Data.Applications {
+			// If the overlay map points to an empty object, delete
+			// it from the base bundle
+			if isZero(reflect.ValueOf(srcAppSpec)) {
+				delete(base.Data.Applications, srcAppName)
+				base.Data.Relations = removeRelations(base.Data.Relations, srcAppName)
+				continue
+			}
+
+			// if this is a new application just append it; otherwise
+			// recursively merge the two application specs.
+			dstAppSpec, defined := base.Data.Applications[srcAppName]
+			if !defined {
+				base.Data.Applications[srcAppName] = srcAppSpec
+				continue
+			}
+
+			mergeStructs(dstAppSpec, srcAppSpec, fpm.forField(srcAppName))
+		}
+	}
+
+	// Merge SAAS blocks
+	if len(overlay.Data.Saas) != 0 {
+		if base.Data.Saas == nil {
+			base.Data.Saas = make(map[string]*SaasSpec, len(overlay.Data.Saas))
+		}
+
+		fpm := overlay.PresenceMap.forField("saas")
+		for srcSaasName, srcSaasSpec := range overlay.Data.Saas {
+			// If the overlay map points to an empty object, delete
+			// it from the base bundle
+			if isZero(reflect.ValueOf(srcSaasSpec)) {
+				delete(base.Data.Saas, srcSaasName)
+				continue
+			}
+
+			// if this is a new saas block just append it; otherwise
+			// recursively merge the two saas specs.
+			dstSaasSpec, defined := base.Data.Saas[srcSaasName]
+			if !defined {
+				base.Data.Saas[srcSaasName] = srcSaasSpec
+				continue
+			}
+
+			mergeStructs(dstSaasSpec, srcSaasSpec, fpm.forField(srcSaasName))
```

Comment:
We should probably be including a step that means if you define both an Application named 'foo' and a SAAS named 'foo' then the bundle is incorrect, since I'm pretty sure you can't have both in a model.

Reply:
This case is handled when the bundle is verified: https://github.com/juju/charm/blob/v6/bundledata.go#L542